e2e/e2e_partition_test.go

```diff
@@ -1,27 +1,132 @@
 package e2e
 
 import (
+	"fmt"
+	"math"
+	"math/rand"
+	"strconv"
 	"testing"
 	"time"
+
+	"github.com/0xPolygon/pbft-consensus"
+	"github.com/stretchr/testify/assert"
 )
 
 func TestE2E_Partition_OneMajority(t *testing.T) {
+	const nodesCnt = 5
 	hook := newPartitionTransport(300 * time.Millisecond)
 
-	c := newPBFTCluster(t, "majority_partition", "prt", 5, hook)
+	c := newPBFTCluster(t, "majority_partition", "prt", nodesCnt, hook)
 	c.Start()
+	defer c.Stop()
 
-	c.WaitForHeight(5, 1*time.Minute)
+	err := c.WaitForHeight(5, 1*time.Minute)
+	assert.NoError(t, err, "Error should not be returned")
 
 	// create two partitions.
-	hook.Partition([]string{"prt_0", "prt_1", "prt_2"}, []string{"prt_3", "prt_4"})
+	majorityPartition := []string{"prt_0", "prt_1", "prt_2"}
+	minorityPartition := []string{"prt_3", "prt_4"}
+	hook.Partition(majorityPartition, minorityPartition)
 
 	// only the majority partition will be able to sync
-	c.WaitForHeight(10, 1*time.Minute, []string{"prt_0", "prt_1", "prt_2"})
+	err = c.WaitForHeight(10, 1*time.Minute, majorityPartition)
+	assert.NoError(t, err, "Error should not be returned")
 
 	// the partition with two nodes is stuck
-	c.IsStuck(10*time.Second, []string{"prt_3", "prt_4"})
+	c.IsStuck(10*time.Second, minorityPartition)
+
+	// reset all partitions
+	hook.Reset()
+
+	allNodes := make([]string, len(c.nodes))
+	for i, node := range c.Nodes() {
+		allNodes[i] = node.name
+	}
+	// all nodes should be able to sync
+	err = c.WaitForHeight(15, 1*time.Minute, allNodes)
+	assert.NoError(t, err, "Error should not be returned")
+}
+
+func TestE2E_Partition_MajorityCanValidate(t *testing.T) {
+	const nodesCnt = 7 // N=3F + 1, F = 2
+	hook := newPartitionTransport(300 * time.Millisecond)
+
+	c := newPBFTCluster(t, "majority_partition", "prt", nodesCnt, hook)
+	limit := int(math.Floor(nodesCnt*2.0/3.0)) + 1 // 2F+1 nodes can Validate
+	for _, node := range c.Nodes() {
+		node.StartWithBackendFactory(createBackend(node.name >= "prt_"+strconv.Itoa(limit)))
+	}
+	defer c.Stop()
+	names := generateNodeNames(0, limit, "prt_")
+	err := c.WaitForHeight(4, 1*time.Minute, names)
+	assert.NoError(t, err, "Error should not be returned.")
+	// restart minority and wait to sync
+	for _, node := range c.Nodes() {
+		if node.name >= "prt_"+strconv.Itoa(limit) {
+			node.Restart()
+		}
+	}
+	err = c.WaitForHeight(4, 1*time.Minute)
+	assert.NoError(t, err, "Error should not be returned.")
+}
+
+func TestE2E_Partition_MajorityCantValidate(t *testing.T) {
+	const nodesCnt = 7 // N=3F + 1, F = 2
+	hook := newPartitionTransport(300 * time.Millisecond)
+
+	c := newPBFTCluster(t, "majority_partition", "prt", nodesCnt, hook)
+	limit := int(math.Floor(nodesCnt * 2.0 / 3.0)) // + 1 removed because 2F+1 nodes is majority
+	for _, node := range c.Nodes() {
+		node.StartWithBackendFactory(createBackend(node.name < "prt_"+strconv.Itoa(limit)))
```

Comment:
Ok, so you are using a `faulty` backend with `createBackend` to simulate that only a majority of nodes are available? If the backend is faulty, no blocks are validated. I like that we support this but the API to introduce this error is different from the partition one.

For partitions we do: 
```
hook.Partition([]string{}...)
```
And to drop a node we do:
```
c.StopNode("ptr_0")
```
The errors are introduced with some high level API either from the hooks (communication transport) or the cluster object (node errors).

The standard way to do this would be something like this:
```
func (c *cluster) FailNode(name string) {
  c.nodes[name].setValidationFails(true)
}

type fsm struct {
  validationFails bool
}

func (f *fsm) setValidationFails(v bool) {
  f.validationFails = v
}

func (f *fsm) Validate(proposal []byte) error {
  if f.validationFails {
    return fmt.Errorf("error")
  }
  return nil
}
```

Reply:
Changes made accordingly.